Additional tests for experiment API

harvey-intel opened the pull request.

ericlee0803 reviewed (changes requested):
test_experiment_with_custom_loop sometimes doesn't pass. 

For example, this one failed: 
https://app.sigopt.com/experiment/464902
https://app.sigopt.com/run/133643

The precise error is as follows:
```
    def log_default_params(self):
>     all_xgb_params = get_all_run_params(self.model, num_boost_round=self.num_boost_round, **self.params)
E     TypeError: get_all_run_params() got multiple values for argument 'booster'

sigopt/xgboost/run.py:178: TypeError
```

I think it's because there's overlap in the run_params drawn from _form_random_run_params and the parameters list. Recommend removing eta which is the overlapping guy. 

This begs the question: right now we are checking that there is no overlap at the experiment level. Should there be a check at the runs level as well?

harvey-intel commented on integration_test/test_xgboost_experiment.py:
```diff
@@ -78,13 +78,30 @@ def _form_random_experiment_config(self, task):
+      for param in parameters_list:
+        if param.type in ('double', 'int'):
+          assert param.bounds.min <= ssr.assignments[param.name] <= param.bounds.max
+        else:
+          categories = [c.name for c in param.categorical_values]
+          assert ssr.assignments[param.name] in categories
```
grid parameter is not a type, it's either "double" or "int".

ericlee0803 replied:
ohhh right

harvey-intel commented:
> test_experiment_with_custom_loop sometimes doesn't pass.
> 
> For example, this one failed: https://app.sigopt.com/experiment/464902 https://app.sigopt.com/run/133643
> 
> The precise error is as follows:
> 
> ```
>     def log_default_params(self):
> >     all_xgb_params = get_all_run_params(self.model, num_boost_round=self.num_boost_round, **self.params)
> E     TypeError: get_all_run_params() got multiple values for argument 'booster'
> 
> sigopt/xgboost/run.py:178: TypeError
> ```
> 
> This is really weird. If I change the following code snippet from utils.py:
> 
> ```
> def get_all_run_params(booster, **train_params):
>   """
>   Get all parameters of a run.
> 
>   Parameters
>   ----------
>   booster : xgboost.Booster
>       Booster model.
>   train_kwargs:
>       Train args already set.
> 
>   Returns
>   ----------
>       dict of run parameters
>   """
>   params = get_train_params(**train_params)
>   params.update(get_booster_params(booster))
>   return params
> ```
> 
> to use the variable "bst" instead of the variable "booster", the test passes (note that I fixed the randomness so it ought to not be the randomness. Alternatively, if I remove the "booster" variable from the parameter search space in the experiment_config, the test also passes (I also changed to use tree_method and I get a pass as well). Is there some kind of namespace overloading that is causing a problem here?
> 
> Instructions to reproduce my error: in test_xgboost_run.py, simply force _form_random_run_params to use the full set of parameters (I did this by setting subset_keys to be POSSIBLE_PARAMETERS.keys() in line 103)

I think I forgot to rebase...which caused this error. I've seen it before.
